### PR TITLE
Support match_phrase filter function in SQL and PPL

### PR DESCRIPTION
**Pull request by MaxKsyunz**
### Description
1. Support match_phrase filter function in SQL.
2. Support match_phrase filter function in PPL.
 
### Issues Resolved
Resolves #185 except for backwards compatibility with the legacy SQL engine. This will be addressed in a separate commit.
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass, including unit test, integration test and doctest
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
  - [x] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: core/src/main/java/org/opensearch/sql/expression/function/OpenSearchFunctions.java [hunk 1]
```diff
@@ -25,54 +25,38 @@
...
   public void register(BuiltinFunctionRepository repository) {
     repository.register(match());
+    repository.register(match_phrase());
   }
 
   private static FunctionResolver match() {
     FunctionName funcName = BuiltinFunctionName.MATCH.getName();
+    // At least field and query parameters
+    // At most field, query, and all optional parameters
+    return getFunctionResolver(funcName, 2, 14);
```
Comment (acarbonetto) [inline core/src/main/java/org/opensearch/sql/expression/function/OpenSearchFunctions.java:35]: magic numbers

File: core/src/main/java/org/opensearch/sql/expression/function/OpenSearchFunctions.java [hunk 2]
```diff
@@ -25,54 +25,38 @@
...
+  }
+
+  private static FunctionResolver match_phrase() {
+    FunctionName funcName = BuiltinFunctionName.MATCH_PHRASE.getName();
+    // At least field and query parameters
+    // At most field, query, and all optional parameters
+    return getFunctionResolver(funcName, 2, 5);
```
Comment (acarbonetto) [inline core/src/main/java/org/opensearch/sql/expression/function/OpenSearchFunctions.java:42]: magic numbers

File: opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/MatchPhraseQuery.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+import org.opensearch.index.query.QueryBuilder;
+import org.opensearch.index.query.QueryBuilders;
+import org.opensearch.sql.data.model.ExprValue;
+import org.opensearch.sql.exception.SemanticCheckException;
+import org.opensearch.sql.expression.Expression;
+import org.opensearch.sql.expression.FunctionExpression;
+import org.opensearch.sql.expression.NamedArgumentExpression;
+import org.opensearch.sql.opensearch.storage.script.filter.lucene.LuceneQuery;
+
+public class MatchPhraseQuery extends LuceneQuery {
```
Comment (acarbonetto) [inline opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/MatchPhraseQuery.java:21]: javadoc would be nice

File: opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/MatchPhraseQuery.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+  ImmutableMap<Object, Object> argAction = ImmutableMap.builder()
+          .put("analyzer", analyzer)
+          .put("slop", slop)
+          .put("zero_terms_query", zeroTermsQuery)
+          .build();
+
+  @Override
+  public QueryBuilder build(FunctionExpression func) {
+    Iterator<Expression> iterator = func.getArguments().iterator();
+    NamedArgumentExpression field = (NamedArgumentExpression) iterator.next();
```
Comment (acarbonetto) [inline opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/MatchPhraseQuery.java:39]: You should either handle (try..catch) the potential NoSuchElementException, OR check hasNext(), 
and return a SemanticCheckException instead. 
We should have unit tests for this

File: opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/MatchPhraseQuery.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    NamedArgumentExpression query = (NamedArgumentExpression) iterator.next();
```
Comment (acarbonetto) [inline opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/MatchPhraseQuery.java:40]: ditto

File: opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/MatchPhraseQuery.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    MatchPhraseQueryBuilder queryBuilder = QueryBuilders.matchPhraseQuery(
+            field.getValue().valueOf(null).stringValue(),
+            query.getValue().valueOf(null).stringValue());
+    while (iterator.hasNext()) {
+      NamedArgumentExpression arg = (NamedArgumentExpression) iterator.next();
```
Comment (acarbonetto) [inline opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/MatchPhraseQuery.java:45]: ditto

File: integ-test/src/test/java/org/opensearch/sql/ppl/WhereCommandIT.java [hunk 7]
```diff
@@ -6,14 +6,15 @@
 
 package org.opensearch.sql.ppl;
 
-import static org.opensearch.sql.legacy.TestsConstants.TEST_INDEX_ACCOUNT;
-import static org.opensearch.sql.legacy.TestsConstants.TEST_INDEX_BANK;
-import static org.opensearch.sql.legacy.TestsConstants.TEST_INDEX_BANK_WITH_NULL_VALUES;
+import static org.opensearch.sql.legacy.TestsConstants.*;
```
Comment (Yury-Fridlyand) [inline integ-test/src/test/java/org/opensearch/sql/ppl/WhereCommandIT.java:9]: CheckStyle could reject wildcard import.

File: integ-test/src/test/java/org/opensearch/sql/ppl/WhereCommandIT.java [hunk 8]
```diff
@@ -110,4 +112,22 @@ public void testRelevanceFunction() throws IOException {
...
+  }
+
+  @Test
+  public void testMathPhraseWithSlop() throws IOException {
+    JSONObject result =
+            executeQuery(
+                    String.format(
+                            "source=%s | where match_phrase(firstname, 'Hattie', slop = 2) | fields firstname", TEST_INDEX_BANK));
+    verifyDataRows(result, rows("Hattie"));
+  }
```
Comment (Yury-Fridlyand) [inline integ-test/src/test/java/org/opensearch/sql/ppl/WhereCommandIT.java:132]: Test should actually use phrases for search, otherwise we can't verify that `match_phrase` works as expected.

Comment (MaxKsyunz) [reply integ-test/src/test/java/org/opensearch/sql/ppl/WhereCommandIT.java]: Fixed. See [7c3ff01](https://github.com/Bit-Quill/opensearch-project-sql/pull/48/commits/7c3ff011ef2e3fd6a7c81df508d3e6dadb6288eb)

File: integ-test/src/test/java/org/opensearch/sql/ppl/WhereCommandIT.java (hunk 7 above)
Comment (MaxKsyunz) [reply integ-test/src/test/java/org/opensearch/sql/ppl/WhereCommandIT.java]: Done. See [d2a45cc](https://github.com/Bit-Quill/opensearch-project-sql/pull/48/commits/d2a45cccee013d1a4f36bab6e1899560135278ad)

File: core/src/main/java/org/opensearch/sql/expression/function/OpenSearchFunctions.java (hunk 1 above)
Comment (MaxKsyunz) [reply core/src/main/java/org/opensearch/sql/expression/function/OpenSearchFunctions.java]: Fixed. See [8d7039a](https://github.com/Bit-Quill/opensearch-project-sql/pull/48/commits/8d7039a2412281f337b2cff7de4ce8204d7e30a0)

File: core/src/main/java/org/opensearch/sql/expression/function/OpenSearchFunctions.java (hunk 2 above)
Comment (MaxKsyunz) [reply core/src/main/java/org/opensearch/sql/expression/function/OpenSearchFunctions.java]: Resolved. See [8d7039a](https://github.com/Bit-Quill/opensearch-project-sql/pull/48/commits/8d7039a2412281f337b2cff7de4ce8204d7e30a0)

File: opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/MatchPhraseQuery.java (hunk 3 above)
Comment (MaxKsyunz) [reply opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/MatchPhraseQuery.java:26]: Done. See https://github.com/Bit-Quill/opensearch-project-sql/commit/d3bd5df03ee25f4aa073bfbbc451724c65ccb504

File: opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/MatchPhraseQuery.java (hunk 6 above)
Comment (MaxKsyunz) [reply opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/MatchPhraseQuery.java:61]: I've updated this function so that next is either not called (see lines 53 and 54) or is preceded by hasNext check like in this loop.

File: opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/MatchPhraseQuery.java (hunk 5 above)
Comment (MaxKsyunz) [reply opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/MatchPhraseQuery.java]: I've updated this function so that next is either not called (see lines 53 and 54) or is preceded by hasNext check like in this loop. See https://github.com/Bit-Quill/opensearch-project-sql/commit/d3bd5df03ee25f4aa073bfbbc451724c65ccb504

File: opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/MatchPhraseQuery.java (hunk 4 above)
Comment (MaxKsyunz) [reply opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/MatchPhraseQuery.java]: Fixed. See https://github.com/Bit-Quill/opensearch-project-sql/commit/d3bd5df03ee25f4aa073bfbbc451724c65ccb504

File: core/src/test/java/org/opensearch/sql/expression/function/OpenSearchFunctionsTest.java [hunk 9]
```diff
@@ -98,16 +105,34 @@ void match() {
...
+    }
+  }
+
+
+  List<FunctionExpression> match_phrase_dsl_expressions() {
+    return List.of(
+      dsl.match_phrase(field, query),
+      dsl.match_phrase(field, query, analyzer),
+      dsl.match_phrase(field, query, analyzer, zeroTermsQueryAll),
+      dsl.match_phrase(field, query, analyzer, zeroTermsQueryNone, slop)
```
[the remark this answers is not in the document]
Comment (MaxKsyunz) [reply core/src/test/java/org/opensearch/sql/expression/function/OpenSearchFunctionsTest.java:132]: That's not necessary, especially since at this level we do not validate argument names. Do you agree?

File: core/src/main/java/org/opensearch/sql/expression/function/OpenSearchFunctions.java [hunk 10]
```diff
@@ -25,54 +25,39 @@
 public class OpenSearchFunctions {
   public void register(BuiltinFunctionRepository repository) {
     repository.register(match());
+    repository.register(match_phrase());
   }
 
   private static FunctionResolver match() {
     FunctionName funcName = BuiltinFunctionName.MATCH.getName();
+    // At most field, query, and all optional parameters
+    final int matchMaxNumParameters = 14;
```
Comment (acarbonetto) [inline core/src/main/java/org/opensearch/sql/expression/function/OpenSearchFunctions.java:34]: if you move these into a public static final variable, it can be imported into other files. 
At least the least, we could then write tests to test these boundary cases using an imported value (like error scenarios where the max number of parameters is exceeeded).

File: core/src/main/java/org/opensearch/sql/expression/function/OpenSearchFunctions.java [hunk 11]
```diff
@@ -25,54 +25,39 @@
...
+    return getRelevanceFunctionResolver(funcName, matchMaxNumParameters);
+  }
+
+  private static FunctionResolver match_phrase() {
+    FunctionName funcName = BuiltinFunctionName.MATCH_PHRASE.getName();
+    // At most field, query, and all optional parameters
+    final int matchPhraseMaxNumParameters = 5;
```
Comment (acarbonetto) [inline core/src/main/java/org/opensearch/sql/expression/function/OpenSearchFunctions.java:41]: ditto

**Review by acarbonetto — APPROVED** [review]

File: core/src/main/java/org/opensearch/sql/expression/function/OpenSearchFunctions.java (hunk 10 above)
Comment (MaxKsyunz) [reply core/src/main/java/org/opensearch/sql/expression/function/OpenSearchFunctions.java]: Addressed in https://github.com/Bit-Quill/opensearch-project-sql/pull/48/commits/e6f7e2fc36f13e9904dab9ef106c49eec89ef25d

File: core/src/main/java/org/opensearch/sql/expression/function/OpenSearchFunctions.java (hunk 11 above)
Comment (MaxKsyunz) [reply core/src/main/java/org/opensearch/sql/expression/function/OpenSearchFunctions.java]: addressed in https://github.com/Bit-Quill/opensearch-project-sql/pull/48/commits/e6f7e2fc36f13e9904dab9ef106c49eec89ef25d

File: core/src/test/java/org/opensearch/sql/expression/function/OpenSearchFunctionsTest.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (Yury-Fridlyand) [reply core/src/test/java/org/opensearch/sql/expression/function/OpenSearchFunctionsTest.java:132]: Yes, thank you.

**Review by Yury-Fridlyand — APPROVED** [review]

**Review by forestmvey — APPROVED** [review]
